enhancements/machine-api/spot-instances.md

```diff
@@ -0,0 +1,459 @@
+Similar to AWS, we can make an optional struct for SpotVMOptions, which, if present, implies the priority is `Spot`.
+
+```
+type SpotVMOptions struct {
+  MaxPrice *float64 `json:”maxPrice,omitempty”`
+}
+
+type AzureMachineProviderSpec struct {
+  ...
+
+  SpotVMOptions *SpotVMOptions `json:”spotVMOptions,omitempty”`
+}
+```
+
+###### Deallocation
+Since Spot VMs are not deleted when they are preempted and instead are deallocated,
+we will need to monitor for deallocated instances and manually cleanup the leftover resources (VM, disks, networking).
+A MachineHealthCheck on Spot VMs should be capable of noticing that the machine has stopped and initiating a delete on the VM by the Machine actuator.
+
+### Risks and Mitigations
+
+#### Control-Plane instances
+
+At present, if a control-plane node is terminated, manual intervention is required to replace the node.
+Therefore, running control-plane machines on top of spot instances,
+where the likelihood of being terminated is increased, should be forbidden.
+
+This risk will be documented and it will be strongly advised that users do not attempt to create control-plane instances on spot instances.
+
+#### Spot instances and Autoscaling
+
+The Kubernetes Cluster Autoscaler, deployed to Openshift clusters, is currently unaware of differences
+between Spot instances and on-demand instances.
+
+If, while the cloud provider has no capacity, or the bid price is too low for AWS/Azure,
+the autoscaler were to try and scale a MachineSet backed by spot instances,
+new Machines would be created, but instances on the cloud provider would not manifest and the Machines would enter a `Failed` state.
+If a MachineHealthCheck were deployed, these Machines would be considered unhealthy and would be deleted, creating a new Machine in its place.
+
+With or without an MHC, the scale request would never manifest in new compute capacity joining the cluster.
+After a half hour period, the autoscaler considers the MachineSet to be out of sync and rescales it to remove the `Failed` instances.
+At this point, it deems the original unschedulable Pods as unschedulable again and may attempt to rescale the same MachineSet.
+
+Based on the [working implementation](https://github.com/kubernetes/autoscaler/pull/2235/files) of Spot instances in the AWS autoscaler provider,
+if the autoscaler Machine API implementation were to provide fake provider IDs for Machines that have failed ([example](https://github.com/JoelSpeed/autoscaler/commit/11ebd1ffdadebbb20d2fac9aae30646b4f47dfa9)),
+the autoscaler would deem these Machines as having unregistered nodes and, after a 15 minute period,
+would request these unregistered nodes be deleted, mark the MachineSet as unhealthy and attempt to scale an alternative MachineSet.
+This, while still not perfect, is preferable to the current state of the autoscaler;
+assuming there is a on-demand based MachineSet to fall back to, this would be tried as a backup.
```

Comment:
may be add a note this is actually current behaviour for aws provider as well https://github.com/kubernetes/autoscaler/pull/2235